pkg/apiserver/database/gorm/provider.go

```diff
@@ -0,0 +1,276 @@
+		return models.Provider{}, err
+	}
+
+	provider.Revision = bumpRevision(dbProvider.Revision)
+
+	existingProvider, err := t.checkUniqueness(provider)
+	if err != nil {
+		var conflictErr *common.ConflictError
+		if errors.As(err, &conflictErr) {
+			return *existingProvider, err
+		}
+		return models.Provider{}, fmt.Errorf("failed to check existing provider: %w", err)
+	}
+
+	marshaled, err := json.Marshal(provider)
+	if err != nil {
+		return models.Provider{}, fmt.Errorf("failed to convert API model to DB model: %w", err)
+	}
+
+	dbObj.Data = marshaled
+
+	if err = t.DB.Save(&dbObj).Error; err != nil {
+		return models.Provider{}, fmt.Errorf("failed to save provider in db: %w", err)
+	}
+
+	return provider, nil
+}
+
+// nolint:cyclop
+func (t *ProvidersTableHandler) UpdateProvider(provider models.Provider, params models.PatchProvidersProviderIDParams) (models.Provider, error) {
+	if provider.Id == nil || *provider.Id == "" {
+		return models.Provider{}, fmt.Errorf("ID is required to update provider in DB")
+	}
+
+	var dbObj Provider
+	if err := getExistingObjByID(t.DB, providerSchemaName, *provider.Id, &dbObj); err != nil {
+		return models.Provider{}, err
+	}
+
+	var err error
+	var dbProvider models.Provider
+	err = json.Unmarshal(dbObj.Data, &dbProvider)
+	if err != nil {
+		return models.Provider{}, fmt.Errorf("failed to convert DB model to API model: %w", err)
+	}
+
+	if err := checkRevisionEtag(params.IfMatch, dbProvider.Revision); err != nil {
+		return models.Provider{}, err
+	}
+
+	provider.Revision = bumpRevision(dbProvider.Revision)
+
+	dbObj.Data, err = patchObject(dbObj.Data, provider)
+	if err != nil {
+		return models.Provider{}, fmt.Errorf("failed to apply patch: %w", err)
+	}
+
+	var ret models.Provider
+	err = json.Unmarshal(dbObj.Data, &ret)
+	if err != nil {
+		return models.Provider{}, fmt.Errorf("failed to convert DB model to API model: %w", err)
+	}
+
+	existingProvider, err := t.checkUniqueness(ret)
+	if err != nil {
+		var conflictErr *common.ConflictError
+		if errors.As(err, &conflictErr) {
+			return *existingProvider, err
+		}
+		return models.Provider{}, fmt.Errorf("failed to check existing provider: %w", err)
+	}
+
+	if err := t.DB.Save(&dbObj).Error; err != nil {
+		return models.Provider{}, fmt.Errorf("failed to save provider in db: %w", err)
+	}
+
+	return ret, nil
+}
+
+func (t *ProvidersTableHandler) DeleteProvider(providerID models.ProviderID) error {
+	if err := deleteObjByID(t.DB, providerID, &Provider{}); err != nil {
+		return fmt.Errorf("failed to delete provider: %w", err)
+	}
+
+	return nil
+}
+
+func (t *ProvidersTableHandler) checkUniqueness(provider models.Provider) (*models.Provider, error) {
+	var providers []Provider
+	// In the case of creating or updating an provider, needs to be checked whether other provider exists with same properties.
+	// TODO(paralta) Confirm uniqueness criteria
+	filter := fmt.Sprintf("id ne '%s' and displayName eq '%s'", *provider.Id, *provider.DisplayName)
```

Comment:
not sure we need to enforce displayName to be unique. I expected displayName to be an optional field and this criteria will fail if multiple Providers have empty DisplayName.

This code will also panic with a nil pointer exception if DisplayName is not set in the POST request so we need to either enforce DisplayName is set or handle == nil.

Reply:
good point on the second part!

regarding the first part, any thoughts/alternatives on checking provider uniqueness or should we just use provider id?

Reply:
The provider ID UUID is the only field that needs to be unique and that isn't currently externally controllable. Although that might change depending on the registration process so we can probably skip the uniqueness checks for now, wdyt? @chrisgacsal maybe you have some thoughts here?